tools/flash_shelly.py

```diff
@@ -0,0 +1,499 @@
+      fp.close()
+    except (urllib.error.HTTPError, urllib.error.URLError) as err:
+      logger.info("Error: %s" % err)
+      n += 1
+    if info:
+      n=41
+    sleep(1)
+
+  if mode == 'homekit':
+    onlinecheck = info['version']
+  else:
+    onlinecheck = info['fw'].split('/v')[1].split('@')[0]
+
+  if onlinecheck == lfw:
+    print(GREEN + "Successfully flashed %s to %s\033[0m" % (host, lfw))
+    if mode == 'homekit':
+      remove('shelly-flash.zip')
+    else:
+      if info['type'] == "SHRGBW2":
+        print("\nTo finalise flash process you will need to switch 'Modes' in the device WebUI,")
+        print(WHITE + "WARNING!!" + NC + "If you are using this device in conjunction with Homebridge it will")
+        print("result in ALL scenes / automations to be removed within HomeKit.")
+        print("Goto http://$device in your web browser")
+        print("Goto settings section")
+        print("Goto 'Device Type' and switch modes")
+        print("Once mode has been changed, you can switch it back to your preferred mode.")
+  else:
+    print(RED + "Failed to flash %s to %s\033[0m" % (host, lfw))
+    print("Current: %s" % onlinecheck)
+
+def probe_info(device, action, dry_run, silent_run, mode, exclude, exclude_device, forced_version, ffw):
+  flash = False
+  info = None           # firmware versions info
+  model = None          # device model
+  lfw = None            # latest firmware available
+  cfw = None            # current firmware on device
+  cfw_type = 'homekit'  # current firmware type
+  host = device
+  host = host.replace('.local','')
+
+  for i, item in enumerate(exclude_device):
+    exclude_device[i] = exclude_device[i].replace('.local','')
+
+  logger.info("\n" + WHITE + "probe_info" + NC)
+  logger.info("device: %s" % device)
+  logger.info("host: %s" % host)
+  logger.info("action: %s" % action)
+  logger.info("dry_run: %s" % dry_run)
+  logger.info("silent_run: %s" % silent_run)
+  logger.info("mode: %s" % mode)
+  logger.info("exclude: %s" % exclude)
+  logger.info("exclude_device: %s" % exclude_device)
+  logger.info("forced_version: %s" % forced_version)
+  logger.info("ffw: %s" % ffw)
+
+  try:
+    fp = urllib.request.urlopen('http://%s/rpc/Shelly.GetInfo' % device)
+    info = json.load(fp)
+    fp.close()
+  except (urllib.error.HTTPError) as err:
+    try:
+      cfw_type = 'stock'
+      fp = urllib.request.urlopen('http://%s/Shelly.GetInfo' % device)
+      info = json.load(fp)
+      fp.close()
+    except (urllib.error.HTTPError, urllib.error.URLError) as err:
+      return 1
+  except (urllib.error.URLError) as err:
+    logger.warning("Could not resolve host: %s" % host)
+    return 1
+
+  if mode == 'keep':
+    mode = cfw_type
+
+  if cfw_type == 'homekit':
+    type = info['app']
+    cfw = info['version']
+    if mode == 'homekit':
+      model = info['model'] if 'model' in info else shelly_model(type, mode)
+      for i in homekit_release_info:
+        if re.search(i[0][0], cfw):
```

Comment:
shouldn't this be i[0]? i think i[0][0] is the first character of the regex

Reply:
nope.

Reply:
oh you where correct close.
```
re.search(i[0][0]: <re.Match object; span=(0, 1), match='2'>
```
I'll tweak to
```
re.search(i[0]: <re.Match object; span=(0, 5), match='2.2.1'>
```

Reply:
done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/2abe8439bc41b318b9192a0ee5c6aaacd23340c8